doc/contribute/guidelines.rst

```diff
@@ -830,14 +830,23 @@ If the change addresses a GitHub issue, include a line of the form:
 
 .. code-block:: none
 
-   Fixes #[issue number]
+   Fixes: https://github.com/zephyrproject-rtos/zephyr/issues/[issue number]
 
 Where ``[issue number]`` is the relevant GitHub issue's number. For
 example:
 
 .. code-block:: none
 
-   Fixes: #1234
+   Fixes: https://github.com/zephyrproject-rtos/zephyr/issues/1234
+
+.. note::
+   References to issue or PRs in GitHub or any other bug tracking system in the
+   commit message must use a full URL rather than just the reference number.
+   This is because the number alone is meaningless in any fork of the
+   repository other than the one in the zephyrproject-rtos org.
```

Comment:
If this is the rationale given, it would make more sense to use the short form (`zephyrproject-rtos/zephyr#1234`) than the full URL, as already mentioned above.

As for being able to click the link and get to the GitHub issue hassle-free from the terminal git log, that is a whole other story and requires further discussion ... IMHO, it is not necessarily a bad thing apart from the "clutter" it creates.

But, a more fundamental question we should be asking is: is this something we should be recommending/encouraging to the newcomers? In general, directly referencing GitHub issues in a commit message in place of describing the related problem should be avoided because it requires one to follow an external link, read up on the issue, and also potentially go through lengthy comments on the issue to understand the full context.

A more "correct" recommendation would be to avoid putting anything GitHub-related in commit messages, especially something like `Fixes #ISSUE` -- that belongs in the PR description on GitHub, not in git commit messages.

Instead, commit messages should contain, if not full, a summary of the related issue(s) alongside a link to the issue if necessary; [for example](https://github.com/zephyrproject-rtos/zephyr/commit/7d5ccc083c05a6c3b4434ed1a2e29fbde56aa76a):

```
cmake: llvm: arm: Remove '--specs' when building with clang for arm
Currently, clang produces a warning that the '--specs=nosys.specs' argument
is unused. Remove this flag to fix the warning as clang will generally not
honor '--specs' when targeting arm baremetal.

While this flag has been present for some time, I think this should be safe
to do. For background, clang does not seem to handle '--specs' [1] besides
possibly passing it through to GCC if GCC is used as the driver for
linking. However, whether GCC will be used for linking as a fallback
depends on the "Toolchain" [1] clang uses internally, which in turn depends
on the triple. For arm/thumb baremetal triples, the Toolchain clang uses
will not fall back to GCC to drive linking, so '--specs' will never be used
here.

I believe this behavior in clang is fairly longstanding as well (since
~2017/LLVM 5 [2]). While there isn't a minimum required clang version for
Zephyr, Zephyr currently requires lld >= 14.0.0. So, I don't think removing
this flag should impact current users (besides preventing the warning).

[1] https://clang.llvm.org/docs/DriverInternals.html
[2] https://reviews.llvm.org/D33259

Signed-off-by: Jonathon Penix <jpenix@quicinc.com>
```

Reply:
> A more "correct" recommendation would be to avoid putting anything GitHub-related in commit messages, especially something like `Fixes #ISSUE` -- that belongs in the PR description on GitHub, not in git commit messages.

Could not agree more, in fact that was my initial intent with this PR: discouraging referring to the issue and instead write something like "use a decent message" and refer to the issue in the PR comment if you want auto-close.

Unfortunately when I started doing that I realized that we were already suggesting to use references in the commit, which to me is very wrong, but I imagine removing that would meet a strong pushback.

This whole thing was in reaction to an existing approved PR that I blocked last second because the whole content of the commit was "Fix issue #xxxx", which I think we agree is unacceptable.